input/listen.go

```diff
@@ -8,126 +8,161 @@ import (
 	"github.com/jpillora/backoff"
 )
 
-func listen(addr string, handler Handler) error {
+type Listener struct {
+	workerPool
+	tcpList *net.TCPListener
+	udpConn *net.UDPConn
+	handler Handler
+}
+
+func NewListener(handler Handler) *Listener {
+	return &Listener{
+		handler: handler,
+	}
+}
+
+func (l *Listener) listen(addr string) error {
+	l.startStoppable()
+
 	// listeners are set up outside of accept* here so they can interrupt startup
-	listener, err := listenTcp(addr)
+	err := l.listenTcp(addr)
 	if err != nil {
 		return err
 	}
 
-	udp_conn, err := listenUdp(addr)
+	err = l.listenUdp(addr)
 	if err != nil {
 		return err
 	}
 
-	go acceptTcp(addr, listener, handler)
-	go acceptUdp(addr, udp_conn, handler)
+	l.wg.Add(2)
+	go l.accept(addr, "tcp", l.acceptTcp, l.listenTcp, l.tcpList)
+	go l.accept(addr, "udp", l.consumeUdp, l.listenUdp, l.udpConn)
 
 	return nil
 }
 
-func listenTcp(addr string) (*net.TCPListener, error) {
-	laddr, err := net.ResolveTCPAddr("tcp", addr)
-	if err != nil {
-		return nil, err
-	}
-	return net.ListenTCP("tcp", laddr)
-}
+func (l *Listener) accept(addr, proto string, consume func(string), reconnect func(string) error, listener Closable) {
+	defer l.wg.Done()
 
-func acceptTcp(addr string, listener *net.TCPListener, handler Handler) {
-	var err error
-	l := listener
 	backoffCounter := &backoff.Backoff{
 		Min: 500 * time.Millisecond,
 		Max: time.Minute,
 	}
 
-	for {
-		log.Notice("listening on %v/tcp", addr)
+	go func() {
+		<-l.shutdown
+		log.Info("shutting down %v/%s, closing socket", addr, proto)
+		listener.Close()
+	}()
 
-		for {
-			// wait for a tcp connection
-			c, err := l.AcceptTCP()
-			if err != nil {
-				log.Error("error accepting on %v/tcp, closing connection: %s", addr, err)
-				l.Close()
-				break
+	for {
+		log.Notice("listening on %v/%s", addr, proto)
+
+		consume(addr)
+
+		select {
+		case <-l.shutdown:
+			return
+		default:
+			for {
```

Comment:
can save a level of indent here by making the default case empty in the select (moving the for loop after the select)